hail/python/hail/experimental/ldscsim.py

```diff
@@ -508,69 +558,116 @@ def calculate_phenotypes(mt, genotype, beta, h2, popstrat=None, popstrat_var=Non
         Column field containing population stratification term.
     popstrat_var : :obj:`float` or :obj:`int`
         Variance of population stratification term.
-        
+    exact_h2: :obj:`bool`
+        Whether to exactly simulate ratio of variance of genetic component of 
+        phenotype to variance of phenotype to be h2. If `False`, ratio will be
+        h2 in expectation. Observed h2 in the simulation will be close to 
+        expected h2 for large-scale simulations.
+
     Returns
     -------
     :class:`.MatrixTable`
         :class:`.MatrixTable` with simulated phenotype as column field.
     """
-    h2 = h2.tolist() if type(h2) is np.ndarray else ([h2] if type(h2) is not list else h2) 
-    assert popstrat_var is None or (popstrat_var >= 0), 'popstrat_var must be non-negative'
-    tid = ''.join(random.choices(string.ascii_uppercase + string.ascii_lowercase, k=5)) # "temporary id" -- random string to identify temporary intermediate fields generated by this method
+    print('calculating phenotype')
+    h2 = h2.tolist() if type(h2) is np.ndarray else (
+        [h2] if type(h2) is not list else h2)
+    assert popstrat_var is None or (
+        popstrat_var >= 0), 'popstrat_var must be non-negative'
+    # tid = "temporary id" -- random string to identify temporary intermediate fields generated by this method
+    tid = _get_tid()
     mt = annotate_all(mt=mt,
-                      row_exprs={'beta_'+tid:beta},
-                      col_exprs={} if popstrat is None else {'popstrat_'+tid:popstrat},
-                      entry_exprs={'gt_'+tid:genotype.n_alt_alleles() if genotype.dtype is dtype('call') else genotype})
+                      row_exprs={'beta_'+tid: beta},
+                      col_exprs={} if popstrat is None else {'popstrat_'+tid: popstrat},
+                      entry_exprs={'gt_'+tid: genotype.n_alt_alleles() if genotype.dtype is dtype('call') else genotype})
+    mt = mt.filter_rows(hl.agg.stats(mt['gt_'+tid]).stdev>0)
     mt = normalize_genotypes(mt['gt_'+tid])
-    if mt['beta_'+tid].dtype == dtype('array<float64>'): #if >1 traits
-        mt = mt.annotate_cols(y_no_noise = hl.agg.array_agg(lambda beta: hl.agg.sum(beta*mt['norm_gt']),mt['beta_'+tid]))
-        mt = mt.annotate_cols(y = mt.y_no_noise + hl.literal(h2).map(lambda x: hl.rand_norm(0,hl.sqrt(1-x))))
+    if mt['beta_'+tid].dtype == dtype('array<float64>'):  # if >1 traits
+        if exact_h2:
+            raise ValueError('exact_h2=True not supported for multitrait simulations')
+            # still working on this
+#            mt = mt.annotate_cols({'y_no_noise_'+tid: hl.agg.array_agg( #use tid because this is the unnormalized-for-h2 version of y_no_noise
```

Comment:
prefer not to have large blocks of commented code -- delete and add implementation later?

Reply:
Yeah no problem, we can delete this for now